### PR TITLE
Fix Nimbus gradle plugin issues [firefox-android: gabrielluong:1818826]

### PR DESCRIPTION
**Pull request by jeddai**
This resolves two issues:
1. The Nimbus gradle plugin will now include the generated files in the sources by default
2. It will now ensure all compilation tasks depend on the nimbus task

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy [hunk 1]
```diff
@@ -363,17 +363,19 @@ class NimbusPlugin implements Plugin<Project> {
...
         def generateSourcesTask = project.tasks.findByName("generate${variant.name.capitalize()}Sources")
         if (generateSourcesTask != null) {
             generateSourcesTask.dependsOn(generateTask)
         } else {
-            project.tasks.findAll().stream().filter({ task ->
-                task.name.startsWith("compile")
-            }).forEach({ task ->
+            project.tasks.findAll().stream()
+            .filter({ task ->
+                return task.name.contains("compile")
```
Comment (jhugman) [inline tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy:376]: Ah I understand why this branch never worked.

File: tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy [hunk 2]
```diff
@@ -363,17 +363,19 @@ class NimbusPlugin implements Plugin<Project> {
             println args
         }
 
-        if(variant.hasProperty('registerJavaGeneratingTask')) {
+        if(variant.metaClass.respondsTo(variant, 'registerJavaGeneratingTask', Task, File)) {
```
Comment (jhugman) [inline tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy:366]: This line was not in the fenix version of the file, so this test has never worked.

**Review by jhugman — APPROVED** [review]
This appears to work fine. 

The theory we have is that the SNAPSHOT subsititution isn't working, but it should be fixed in another ticket. Could you file another ticket to fix that task?

✅  in the meantime. Thanks Charlie.

**Comment by jeddai** [comment]
> The theory we have is that the SNAPSHOT subsititution isn't working, but it should be fixed in another ticket. Could you file another ticket to fix that task?

Yeah so the issue is specifically substitution within the buildscript block. I'll file another ticket and I have some initial findings to share on it

File: tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy (hunk 2 above)
Comment (jeddai) [reply tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy:366]: Yep 😬

File: tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy (hunk 1 above)
Comment (jeddai) [reply tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy:376]: Also yep 😬

**Comment by jeddai** [comment]
Ooooookay so a few things with this PR:
1. The branch build is going to fail because dependencies in the `buildscript` block are not substituted by the a-s code yet. That will be adjusted as a part of [this ticket (EXP-3239)](https://mozilla-hub.atlassian.net/browse/EXP-3239).
2. The swiftlint is (was?) failing for reasons unrelated to my changes

I'm not sure how to override the failing build to merge, any advice there would be great.

File: tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy (hunk 2 above)
Comment (jeddai) [reply tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy:366]: I have very much confirmed it does now though!

**Comment by tarikeshaq** [comment]
> I'm not sure how to override the failing build to merge, any advice there would be great.

You click the `Squash and merge` button, as long a `run-tests` passes (which it does!) that button should be clickable

image


If we're landing this, I'd like to confirm that this **does not** break branch builds, and just breaks branch builds against the fork you're building against? i.e when it comes to other engineers launching branch builds, their branch builds should work fine.. if it's not, that may also be okay, as long as we're aware of which failures are temporary so we can ignore them

**Comment by jeddai** [comment]
Ah okay I’ll check that tomorrow and see if it works, thanks!

No, this does not break branch builds as a whole thankfully. The issue is
with the branch I was building — it is currently failing, and this change
was meant to fix it. Since the plug-in is set up in the buildscript
configuration, and that module substitution isn’t working, that’s why the
branch build is failing.

**Pull request merged**